src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,306 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+//    initializations in run-time (e.g. in Tier1), although, can be enabled for JIT/CG as well.
+//
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+PhaseStatus Compiler::fgExpandStaticInit()
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveStaticInit())
+    {
+        // TP: nothing to expand in the current method
+        JITDUMP("Nothing to expand.\n")
+        return result;
+    }
+
+    if (opts.OptimizationDisabled())
+    {
+        JITDUMP("Optimizations aren't allowed - bail out.\n")
+        return result;
+    }
+
+    // TODO: Replace with opts.compCodeOpt once it's fixed
+    const bool preferSize = opts.jitFlags->IsSet(JitFlags::JIT_FLAG_SIZE_OPT);
+    if (preferSize)
+    {
+        // The optimization comes with a codegen size increase
+        JITDUMP("Optimized for size - bail out.\n")
+        return result;
+    }
+
+    for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+        if (block->isRunRarely())
+        {
+            continue;
+        }
+
+    SCAN_BLOCK_AGAIN:
+        for (Statement* const stmt : block->NonPhiStatements())
+        {
+            if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+            {
+                // TP: Stmt has no calls - bail out
+                continue;
+            }
+
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                bool                    isGc       = false;
+                StaticHelperReturnValue retValKind = {};
+                if (!IsStaticHelperEligibleForExpansion(tree, &isGc, &retValKind))
+                {
+                    continue;
+                }
+
+                GenTreeCall* call = tree->AsCall();
+                if (call->IsTailCall())
+                {
+                    // It is very unlikely, but just in case
+                    continue;
+                }
+
+                JITDUMP("Expanding static initialization for '%s', call: [%06d] in " FMT_BB "\n",
+                        eeGetClassName(call->gtRetClsHnd), dspTreeID(tree), block->bbNum)
+
+                if (call->gtInitClsHnd == NO_CLASS_HANDLE)
+                {
+                    assert(!"helper call was created without gtRetClsHnd");
+                    continue;
+                }
+
+                int                  isInitOffset = 0;
+                CORINFO_CONST_LOOKUP flagAddr     = {};
+                if (!info.compCompHnd->getIsClassInitedFlagAddress(call->gtRetClsHnd, &flagAddr, &isInitOffset))
+                {
+                    JITDUMP("getIsClassInitedFlagAddress returned false - bail out.\n")
+                    continue;
+                }
+
+                CORINFO_CONST_LOOKUP staticBaseAddr = {};
+                if ((retValKind == SHRV_STATIC_BASE_PTR) &&
+                    !info.compCompHnd->getStaticBaseAddress(call->gtRetClsHnd, isGc, &staticBaseAddr))
```

Comment:
If we stored the field handle instead of the owning class, we would not need the `isGc` logic.

Reply:
Ugh, good point, although maybe the current shape is more generic for some future experiments? Let's keep it this way if nobody else minds, JIT-EE modifications drain a lot of energy 🙂